hal/source/gpio_device.cpp

```diff
@@ -0,0 +1,158 @@
+#include "gpio_device.h"
+#include <boost/algorithm/string/classification.hpp>
+#include <boost/algorithm/string/split.hpp>
+
+
+GPIODevice::GPIODevice(NRTFPGADevice & nrt_fpga_device) :
+	nrt_fpga_device_(nrt_fpga_device) {
+	CommonKeyPairs config = {
+	};
+	ReConfig(config);
+}
+
+
+GPIODevice::~GPIODevice() {
+	ClearInitialized();
+}
+
+error_condition GPIODevice::ReConfig(const CommonKeyPairs& config) {
+	ClearInitialized();
+	vector<uint8_t> out_data = {0, 0, 0, 0};
+	nrt_fpga_device_.Set(NRTFPGADevice::kDOData, out_data);
+	SetInitialized();
+
+	return kNoError;
+}
+
+error_condition GPIODevice::Get(const string& key, vector<uint8_t>& result) {
+	if (key.empty())
+		return kInvalidArgument;
+	vector<uint8_t> get_data;
+	if (key == kDIPDataName)
+	{
+		result.resize(2);
+		if (nrt_fpga_device_.Get(NRTFPGADevice::kDIData, get_data) != kNoError)
+			return kInvalidArgument;
+
+		copy(get_data.begin(), get_data.end(), result.begin());
+	}
+	else if (key == kDIHDataName)
+	{
+		result.resize(2);
+		if (nrt_fpga_device_.Get(NRTFPGADevice::kDIData, get_data) != kNoError)
+			return kInvalidArgument;
+
+		copy(get_data.begin() + 2, get_data.end(), result.begin());
+	}
+	else if (key == kDOPDataName)
+	{
+		result.resize(2);
+		if (nrt_fpga_device_.Get(NRTFPGADevice::kDOData, get_data) != kNoError)
+			return kInvalidArgument;
+
+		copy(get_data.begin(), get_data.end(), result.begin());
+	}
+	else if (key == kDOHDataName)
+	{
+		result.resize(2);
+		if (nrt_fpga_device_.Get(NRTFPGADevice::kDOData, get_data) != kNoError)
+			return kInvalidArgument;
+
+		copy(get_data.begin() + 2, get_data.end(), result.begin());
+	}
+	else
+		return kInvalidArgument;
+
+	return kNoError;
+}
+
+error_condition GPIODevice::Get(const string& key, unsigned int& result) {
+	if (key.empty())
+		return kInvalidArgument;
+	int pos = key.find(".", 0);
+	if (pos == -1)
+		return kInvalidArgument;
+	string data_name = key.substr(0, pos);
+	int position = stoi(key.substr(pos + 1));
```

Comment:
Please use `try ... catch`. Please pay attention to the exception section on http://www.cplusplus.com/reference/string/stoi/, if we don't catch an exception, the application will crash.